lib/auth/auth.go

```diff
@@ -1899,6 +1913,115 @@ func (a *Server) GetDatabaseServers(ctx context.Context, namespace string, opts
 	return a.GetCache().GetDatabaseServers(ctx, namespace, opts...)
 }
 
+// mfaAuthChallenge constructs an MFAAuthenticateChallenge for all MFA devices
+// registered by the user.
+func (a *Server) mfaAuthChallenge(ctx context.Context, user string) (*proto.MFAAuthenticateChallenge, error) {
+	// Check what kind of MFA is enabled.
+	apref, err := a.GetAuthPreference()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var enableTOTP, enableU2F bool
+	switch apref.GetType() {
+	case teleport.TOTP:
+		enableTOTP, enableU2F = true, false
+	case teleport.U2F:
+		enableTOTP, enableU2F = false, true
+	default:
+		// Other AuthPreference types don't restrict us to a single MFA type,
+		// send all challenges.
+		enableTOTP, enableU2F = true, true
+	}
+	var u2fPref *types.U2F
+	if enableU2F {
+		u2fPref, err = apref.GetU2F()
+		if trace.IsNotFound(err) {
+			// If U2F parameters were not set in the auth server config,
+			// disable U2F challenges.
+			enableU2F = false
+		} else if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	devs, err := a.GetMFADevices(ctx, user)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	challenge := new(proto.MFAAuthenticateChallenge)
+	for _, dev := range devs {
+		switch dd := dev.Device.(type) {
+		case *types.MFADevice_Totp:
+			if !enableTOTP {
+				continue
+			}
+			challenge.TOTP = new(proto.TOTPChallenge)
+		case *types.MFADevice_U2F:
+			if !enableU2F {
+				continue
+			}
+			ch, err := u2f.AuthenticateInit(ctx, u2f.AuthenticateInitParams{
+				AppConfig:  *u2fPref,
+				Dev:        dd.U2F,
+				Storage:    a.Identity,
+				StorageKey: user,
+			})
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			challenge.U2F = append(challenge.U2F, &proto.U2FChallenge{
+				KeyHandle: ch.KeyHandle,
+				Challenge: ch.Challenge,
+				AppID:     ch.AppID,
+			})
+		default:
+			log.Warningf("skipping MFA device of unknown type %T", dev.Device)
+		}
+	}
+	return challenge, nil
+}
+
+func (a *Server) validateMFAAuthResponse(ctx context.Context, user string, resp *proto.MFAAuthenticateResponse) error {
+	switch res := resp.Response.(type) {
+	case *proto.MFAAuthenticateResponse_TOTP:
+		return a.checkOTP(user, res.TOTP.Code)
+	case *proto.MFAAuthenticateResponse_U2F:
+		return a.checkU2F(ctx, user, res.U2F)
+	default:
+		return trace.BadParameter("unknown or missing MFAAuthenticateResponse type %T", resp.Response)
+	}
+}
+
+func (a *Server) checkU2F(ctx context.Context, user string, res *proto.U2FResponse) error {
+	devs, err := a.GetMFADevices(ctx, user)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, dev := range devs {
```

Comment:
Nice catch!
Performance-wise, it shouldn't be a big problem since most users have at most 2-3 devices.
However, there's no reason to attempt validation against a device when the `KeyHandle` doesn't match. And if it does, it's the only device that can be used for validation.

Updated.